src/backend/distributed/commands/utility_hook.c

```diff
@@ -1627,17 +1664,93 @@ RunPreprocessMainDBCommand(Node *parsetree, const char *queryString)
 static void
 RunPostprocessMainDBCommand(Node *parsetree)
 {
+	if (IsStatementSupportedInNonMainDb(parsetree) &&
+		StatementRequiresMarkDistributedFromNonMainDb(parsetree))
+	{
+		MarkObjectDistributedInNonMainDb(parsetree);
+	}
+}
+
+
+/*
+ * GetObjectInfo returns the name and oid of the object in the given parsetree.
+ */
+static ObjectInfo
+GetObjectInfo(Node *parsetree)
+{
+	ObjectInfo info;
+
 	if (IsA(parsetree, CreateRoleStmt))
 	{
-		StringInfo mainDBQuery = makeStringInfo();
-		CreateRoleStmt *createRoleStmt = castNode(CreateRoleStmt, parsetree);
-		Oid roleOid = get_role_oid(createRoleStmt->role, false);
-		appendStringInfo(mainDBQuery,
-						 MARK_OBJECT_DISTRIBUTED,
-						 AuthIdRelationId,
-						 quote_literal_cstr(createRoleStmt->role),
-						 roleOid,
-						 quote_literal_cstr(CurrentUserName()));
-		RunCitusMainDBQuery(mainDBQuery->data);
+		CreateRoleStmt *stmt = castNode(CreateRoleStmt, parsetree);
+		info.name = stmt->role;
+		info.id = get_role_oid(stmt->role, false);
 	}
+
+	/* Add else if branches for other statement types */
+
+	return info;
+}
+
+
+/*
+ * MarkObjectDistributedInNonMainDb marks the given object as distributed on the
+ * non-main database.
+ */
+static void
+MarkObjectDistributedInNonMainDb(Node *parsetree)
+{
+	ObjectInfo objectInfo = GetObjectInfo(parsetree);
+	StringInfo mainDBQuery = makeStringInfo();
+	appendStringInfo(mainDBQuery,
+					 MARK_OBJECT_DISTRIBUTED,
+					 AuthIdRelationId,
```

Comment:
Can we add .catalogId field into ObjectInfo and let GetObjectInfo() to set .catalogId to AuthIdRelationId for our case?

Reply:
Not relevant since we export the object parameters from statement. We shouldn't add it

Reply:
Fixed as you suggested